app/src/main/java/com/example/sampleappproject/MainActivity.kt

```diff
@@ -2,66 +2,89 @@ package com.example.sampleappproject
 
 import android.os.Bundle
 import android.util.Log
-import android.view.Gravity
-import android.widget.Button
-import android.widget.TextView
 import android.widget.Toast
 import androidx.appcompat.app.AppCompatActivity
+import androidx.core.view.isVisible
 import androidx.lifecycle.ViewModelProvider
+import androidx.paging.CombinedLoadStates
+import androidx.paging.LoadState
 import androidx.recyclerview.widget.RecyclerView
-import com.example.sampleappproject.adapter.Adapter
+import com.example.sampleappproject.databinding.ActivityMainBinding
+import com.example.sampleappproject.paging.CharacterPagingAdapter
+import com.example.sampleappproject.paging.LoaderAdapter
 import com.example.sampleappproject.viewmodel.MainViewModel
 import com.example.sampleappproject.viewmodel.MainViewModelFactory
-
+import kotlinx.coroutines.flow.collect
 
 class MainActivity : AppCompatActivity() {
     private lateinit var mainViewModel: MainViewModel
-    private lateinit var nextButton: Button
-    private lateinit var prevButton: Button
-    private lateinit var pageNumber : TextView
+    private lateinit var mainBinding: ActivityMainBinding
+    private lateinit var recyclerView: RecyclerView
```

Comment:
Why you need **_recyclerView_** as you are already using viewbinding

Reply:
removed it now missed that one.